### PR TITLE
Implement delta on load

### PR DESCRIPTION
**Pull request by lmbarros**
Applies balena deltas when loading them with `balena load`. Follows the same semantics as `balena pull`:

* If the image being loaded (or pulled) is a delta, it is transparently applied.
* The new image (i.e., the one created by applying the delta) will be tagged the same as the delta image.

(I am not particularly fond of these behaviors, but I think it's better to remain consistent. If we ever decide to change them we should do this to both `load` and `pull` at once.)

Some remarks:

* The implementation is based on my previous PoC, with some relevant changes:
    * Refactoring to avoid as much code replication as practical.
    * Better error reporting.
    * More and better automated tests. Now testing the happy path all error paths I could identify.
* Did a round of manual testing, too. Covered the followed cases:
    * Loading a delta image from an exported tar (happy path).
    * Loading a delta image from an exported tar when the base image is not present.
    * Loading a delta image from an exported tar when the target image is already present locally.
    * Did a delta update of a user app (to check for regressions, given the refactoring)
    * Did a delta-based HUP (ditto)

### CODE REVIEW

File: image/tarexport/load.go [hunk 1]
```diff
@@ -101,7 +104,45 @@ func (l *tarexporter) Load(inTar io.ReadCloser, outStream io.Writer, quiet bool)
...
+					return nil
+				}
+
+				config = []byte(targetConfig)
+			}
+		}
+
+		configRootFS, err := imgConfigStore.RootFSFromConfig(config)
+		if err == nil && configRootFS == nil {
+			return errors.New("nil target image root filesystem")
```
Comment (klutchell) [inline image/tarexport/load.go:128]: What is the reasoning for using a mix of `fmt.Errorf`, `outStream.Write`, and `errors.New` in the same function?

File: image/tarexport/load.go [hunk 2]
```diff
@@ -201,15 +242,40 @@ func (l *tarexporter) loadLayer(filename string, rootFS image.RootFS, id string,
 	}
 
 	inflatedLayerData, err := archive.DecompressStream(r)
```
Comment (klutchell) [inline image/tarexport/load.go:233]: How was this working before without importing `archive.tar` as you have added?

File: image/tarexport/load.go [hunk 3]
```diff
@@ -201,15 +242,40 @@ func (l *tarexporter) loadLayer(filename string, rootFS image.RootFS, id string,
...
+
+	layerData := inflatedLayerData
+	if deltaBase != nil {
+		pR, pW := io.Pipe()
+		go func() {
+			tr := tar.NewReader(inflatedLayerData)
```
Comment (klutchell) [inline image/tarexport/load.go]: To answer my above question, I guess `archive` was already imported from elsewhere, and you needed to add tar for this line.

File: image/tarexport/load.go (hunk 1 above)
Comment (lmbarros) [reply image/tarexport/load.go:128]: Interesting question! We use `fmt.Errorf` and `errors.New` to create `error`s and return them. It's the way to say "something wrong happened here and I believe you, dear caller, can handle it better than I can."

We use `outStream.Write`, on the other hand, to say things directly to the human user.

As for why we have two different functions to create `error`s: `fmt.Errorf` simply wraps `errors.New` adding `printf`-like formatting capabilities. It feels inconsistent that they are on different packages, but the Go standard library has placed everything accepting `printf`-like formatting into the `fmt` package. (So... it's consistent in its inconsistency!)

File: image/tarexport/load.go (hunk 3 above)
Comment (lmbarros) [reply image/tarexport/load.go]: Right, exactly this, @klutchell !

File: image/tarexport/load.go (hunk 1 above)
Comment (klutchell) [reply image/tarexport/load.go:128]: Excellent explanation, thank you!

**Comment by lmbarros** [comment]
LGTM

**Pull request merged**